pkg/hub/managedclusterset/controller.go

```diff
@@ -42,34 +53,70 @@ func NewManagedClusterSetController(
 		clusterLister:    clusterInformer.Lister(),
 		clusterSetLister: clusterSetInformer.Lister(),
 		eventRecorder:    recorder.WithComponentSuffix("managed-cluster-set-controller"),
+
+		clusterSetsMap: map[string]string{},
 	}
 	return factory.New().
 		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
 			accessor, _ := meta.Accessor(obj)
 			return accessor.GetName()
 		}, clusterSetInformer.Informer()).
+		// register two event handlers to cluster informer. One haneles the current parent
+		// clusterset, and the other handles the original parent clusterset. The order of
+		// the registering matters. originalClusterSetQueueKeyFunc must be called before
+		// currentClusterSetQueueKeyFunc is called, otherwise clusterSetsMap will be updated
+		// unexpectedly
+		//
 		// TODO, create a PR for library-go to accept a queue key function, which
 		// is able to produce a []string instead of string based on one object, when
 		// registering event handler. And then refactor the logic here.
-		WithInformers(clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(c.currentClusterSetQueueKeyFunc, clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(c.originalClusterSetQueueKeyFunc, clusterInformer.Informer()).
 		WithSync(c.sync).
 		ToController("ManagedClusterSetController", recorder)
 }
 
+// originalClusterSetQueueKeyFunc returns the original clusterset the cluster previously
+// belonged to.
+func (c *managedClusterSetController) originalClusterSetQueueKeyFunc(obj runtime.Object) string {
+	accessor, _ := meta.Accessor(obj)
+	clusterName := accessor.GetName()
+	originalClusterSetName := c.clusterSetsMap[clusterName]
+
+	var currentClusterSetName string
```

Comment:
I think you will need a mutual lock between this func and the one update the map

Reply:
Done.